scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala

```diff
@@ -219,6 +220,8 @@ class SparkWriter(val sparkMode: SparkMode) extends Writer {
       (nextState.copy(id = nextState.id + 1), (nextState.id, acts))
     }
     // now we run the actions:
-    CFuture.uncancellable(Future.traverse(acts)(fn => fn()).map(_ => (id, ExecutionCounters.empty)))
+    CFuture.uncancellable(
+      Future.traverse(acts)(fn => fn()).map(_ => (id, SparkCounters.lazyEvaluateAsExecutionCounters()))
```

Comment:
I am currently confused why getting the SparkCounters state is empty inside of this map scope but if I compute it lazily it has the correct values later when asserting in tests.

Reply:
when placing breakpoints, the map block above hits before the counterPipe is processed in spark

Reply:
@johnynek any tips?

Reply:
These futures complete when spark has returned an RDD in many cases. That may not be when the job is finished.

That's an issue. That hasn't been a problem that I know of so far. So we really need to make sure this Future is only complete when *all* of these writes have fully occurred.

I think the problem is probably with `def force`. So, this part of the future should occur earlier:

https://github.com/twitter/scalding/blob/e44f2f21fa90f669d6bf81af607c5d537f19e93c/scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala#L90

and in the case of `forceToDiskExecution` we need some way for spark to tell us when the caching/materialization is done. I don't know how to do that off the top of my head.

Alternatively, rather than using cache, we could make a temp file ourselves, write to that, which is a spark action and we can block on that, then we can re-read that path ourselves.

Reply:
> These futures complete when spark has returned an RDD in many cases. That may not be when the job is finished.

same conclusion I came to I will look into trying to solve this issue I might also split it off into a separate PR if I figure it out

Reply:
there is a pattern that was historically used inside of spark sql of using the `count` operation to force an RDD computation: 

https://github.com/apache/spark/pull/30403/files#diff-aa467582590608ec702c1caf6208e79af5e6f3ee2cfd58a53bdf2601cd5015d6R65

this is one way to achieve a blocking `cache` operation though it does carry some minimal overhead of re-reading the RDD. 

it's surprising that there's a blocking version of `unpersist` but no blocking `persist` on RDD

Reply:
okay, that makes sense for the Force case, but I think being more direct and fixing the bug with toIterable.